route/option.go

```diff
@@ -232,6 +236,228 @@ func ServiceGroups(serviceGroups []ServiceGroup) Option {
+		stops []sdkRoute.Stop
+		// mutable data
+		Locations     map[string]int `json:"locations,omitempty"`
+		vehicleValues map[string]int
+		fleetValue    int
+	}
+
+The fleetData struct can implement the PlanUpdater interface so that when
+assigning stops to a vehicle, the index of each stop is updated using the
+updated locations of that vehicle's state. In addition to this, a custom value
+function for an assignment is provided by leveraging the custom value that was
+provided in the previous implementation.
+
+	// Track the index of the route for each stop in each vehicle route.
+	// Customize value function to incorporate the custom vehicle engine's
+	// value. Notice that the vehicle state is casted to an Updater interface.
+	// This makes it possible to call the Updater() method and obtain the custom
+	// updated information at the vehicle level, directly from the PlanUpdater.
+	func (d fleetData) Update(
+		plan sdkRoute.PartialPlan,
+		vehicles []sdkRoute.PartialVehicle,
+	) (sdkRoute.PlanUpdater, int, bool) {
+		// Deep copy locations stored.
+		locations := make(map[string]int, len(d.Locations))
+		for stopdID, i := range d.Locations {
+			locations[stopdID] = i
+		}
+		d.Locations = locations
+		// Deep copy the data required for the value function.
+		values := make(map[string]int, len(d.vehicleValues))
+		for vehicleID, i := range d.vehicleValues {
+			values[vehicleID] = i
+		}
+		d.vehicleValues = values
+		for _, vehicle := range vehicles {
+			// Update locations based on the changes made on the vehicle state.
+			vehicleID := vehicle.ID()
+			updater := vehicle.Valuer().(vehicleData)
+			for stopdID, i := range updater.Locations {
+				d.Locations[stopdID] = i
+			}
+
+			// Update value function information.
+			value := vehicle.Update()
+			d.fleetValue -= d.vehicleValues[vehicleID]
+			d.vehicleValues[vehicleID] = value
+			d.fleetValue += d.vehicleValues[vehicleID]
+		}
+		// Remove unassigned locations.
+		for _, location := range plan.Unassigned().Slice() {
+			stop := d.stops[location]
+			delete(d.Locations, stop.ID)
+		}
+		return d, d.fleetValue, true
+	}
+
+*/
+func Update(v VehicleUpdater, f PlanUpdater) Option {
+	connect()
+	return updateFunc(v, f)
+}
+
+// FilterWithRoute adds a new VehicleFilter. Compared to the Filter option, the
+// FilterWithRoute option is more flexible. It defines a function that takes an
+// IntDomain of candidate vehicles, an IntDomain of locations that will be
+// assigned to a particular vehicle, and a slice of routes for all vehicles. It
+// returns an IntDomain representing vehicles that cannot service the domain of
+// locations.
+func FilterWithRoute(
+	filter func(
+		vehicleCandidates model.Domain,
+		locations model.Domain,
+		routes [][]int,
+	) model.Domain,
+) Option {
+	connect()
+	return filterWithRouteFunc(filter)
+}
+
+// Sorter sets the given custom vehicle sorter. The vehicle sorter lets you
+// define a function which returns the vehicle indices in a specific order. The
```

Comment:
Here the word is `indices` and elsewhere it is `indexes` (in `router.go`). Can we settle for one style and apply it throughout?